File: web/src/app/core/context-scan.ts
// The prepass calls: one scan for the file's glossary, then one small
// attribution call per scene that mixes known genders. Both are best-effort —
// on failure the pipeline continues without the hints.

import { SCAN_MAX_TOKENS } from './constants';
import {
  ChatClient,
  ChatOptions,
  ProviderConfig,
  TranslationCancelledError,
  describeError,
  stripMarkdownFences,
  throwIfCancelled,
} from './chat-client';
import { parseAttributionResponse, parseContextResponse } from './context-parse';
import {
  FileContext,
  enrichScenesWithBlockText,
  genderMark,
  needsAttribution,
  recurringPhrases,
  serializeForScan,
  type CharacterHint,
  type SceneHint,
} from './context-pass';
import {
  ATTRIBUTION_SYSTEM_PROMPT,
  CONTEXT_SYSTEM_PROMPT,
  buildAttributionUserMessage,
  buildScanUserMessage,
} from './translation-prompt';
import { CallKind } from './run-stats';
import { SubtitleBlock } from './srt-parser';

/** What the prepass needs from the run's settings. */
export interface ScanOptions extends ChatOptions {
  /** Tallies one LLM call so the run can report what it cost. */
  count: (kind: CallKind) => void;
  dialect: string;
}

export async function extractFileContext(
  chat: ChatClient,
  blocks: SubtitleBlock[],
  sourceLang: string,
  targetLang: string,
  provider: ProviderConfig,
  scanBudget: number,
  run: ScanOptions,
  cancelSignal?: AbortSignal,
): Promise<FileContext> {
  const userMessage = buildScanUserMessage(
    sourceLang, targetLang, serializeForScan(blocks, scanBudget), run.dialect,
    recurringPhrases(blocks),
  );
  try {
    run.count('scan');
    const raw = await chat.callChat(
      CONTEXT_SYSTEM_PROMPT, userMessage, provider, SCAN_MAX_TOKENS, run, cancelSignal,
    );
    const ctx = parseContextResponse(stripMarkdownFences(raw));
    if (ctx.isEmpty()) return ctx;
    return enrichScenesWithBlockText(ctx, blocks);
  } catch (err) {
    if (err instanceof TranslationCancelledError) throw err;
    console.warn('Context scan failed, proceeding without:', describeError(err));
    run.notify('Glossary scan failed - continuing without context hints');
    return new FileContext();
  }
}

export async function refineSceneAttribution(
  chat: ChatClient,
  ctx: FileContext,
  blocks: SubtitleBlock[],
  provider: ProviderConfig,
  concurrency: number,
  run: ScanOptions,
  cancelSignal?: AbortSignal,
  full = false,
  targetInflects = false,
): Promise<void> {
  const genderBy = new Map(
    ctx.characters.map((h) => [h.source.toLowerCase(), h.gender]),
  );
  const targets = ctx.scenes.filter(
    (s) => needsAttribution(s, genderBy, full, targetInflects),
  );
  if (!targets.length) return;

  let nextIdx = 0;
  let failures = 0;
  const worker = async () => {
    while (true) {
      throwIfCancelled(cancelSignal);
      const i = nextIdx++;
      if (i >= targets.length) return;
      const scene = targets[i];
      const prompt = buildSceneAttributionMessage(scene, blocks, ctx.characters);
      if (!prompt) continue;
      try {
        run.count('attribution');
        const raw = await chat.callChat(
          ATTRIBUTION_SYSTEM_PROMPT, prompt.message, provider,
          prompt.lineCount * 20 + 100, run, cancelSignal,
        );
        scene.attribution = parseAttributionResponse(
          stripMarkdownFences(raw), scene, ctx.characters,
        );
      } catch (err) {
        if (err instanceof TranslationCancelledError) throw err;
        failures++;
        console.warn(
          `Attribution failed for blocks ${scene.start}-${scene.end}:`,
          describeError(err),
        );
      }
    }
  };

  const workerCount = Math.min(concurrency, targets.length);
  await Promise.all(Array.from({ length: workerCount }, worker));
  if (failures) {
    run.notify('Speaker attribution failed - continuing without per-block speakers');
  }
}


function buildSceneAttributionMessage(
  scene: SceneHint,
  blocks: SubtitleBlock[],
  characters: CharacterHint[],
): { message: string; lineCount: number } | null {
  const present = new Set(scene.participants);
  const roster = characters
    .filter((h) => present.has(h.source))
    .map((h) => `- ${h.source} (${genderMark(h.gender) || '?'})`)
    .join('\n');
  // Walk the blocks that exist: an inflated range would drive the token budget.
  const sceneLines = blocks
    .filter((b) => b.number >= scene.start && b.number <= scene.end)
    .map((b) => `[${b.number}] ${b.text.replace(/\n/g, ' ')}`);
  if (!roster || !sceneLines.length) return null;
  return {
    message: buildAttributionUserMessage(roster, sceneLines),
    lineCount: sceneLines.length,
  };
}

File: cli/core/context_scan.py
"""The prepass calls: one scan for the file's glossary, then one small
attribution call per scene whose speakers the translation depends on."""

from __future__ import annotations

import asyncio

import httpx

from .chat_client import call_chat_api, strip_markdown_fences
from .config import TranslationConfig
from .constants import SCAN_MAX_TOKENS
from .context_parse import parse_attribution_response, parse_context_response
from .context_pass import (
    CharacterHint,
    FileContext,
    SceneHint,
    clamp_scenes_to_blocks,
    enrich_scenes_with_block_text,
    gender_mark,
    recurring_phrases,
    scenes_needing_attribution,
    serialize_for_scan,
)
from .languages import effective_norms, grammar_note_for
from .prompt import (
    ATTRIBUTION_SYSTEM_PROMPT,
    CONTEXT_SYSTEM_PROMPT,
    build_attribution_user_message,
    build_scan_user_message,
)
from .srt_parser import SubtitleBlock


def attribution_targets(
    context: FileContext, cfg: TranslationConfig,
) -> list[SceneHint]:
    """Scenes worth one attribution call under this run's settings. A target
    whose script inflects for gender qualifies every two-hander scene, known
    genders or not; --full-attribution buys every scene with a cast."""
    norms = effective_norms(cfg.target_lang, cfg.max_line_chars)
    return scenes_needing_attribution(
        context, cfg.full_attribution, bool(grammar_note_for(norms.script)))


async def extract_file_context(
    client: httpx.AsyncClient,
    blocks: list[SubtitleBlock],
    cfg: TranslationConfig,
) -> FileContext:
    """Run one scan call. Returns the parsed+enriched context."""
    user_msg = build_scan_user_message(
        cfg.source_lang, cfg.target_lang,
        serialize_for_scan(blocks, cfg.scan_char_budget),
        cfg.dialect,
        recurring_phrases(blocks),
    )
    try:
        cfg.calls.count("scan")
        raw = await call_chat_api(
            client, CONTEXT_SYSTEM_PROMPT, user_msg, cfg,
            max_tokens=SCAN_MAX_TOKENS,
        )
    except Exception as e:
        cfg.warn(f"    Context scan failed, proceeding without: {e}")
        return FileContext()

    context = parse_context_response(strip_markdown_fences(raw))
    if context.is_empty():
        snippet = (raw or "").strip().replace("\n", " ")[:240]
        cfg.warn(f"    Context scan returned empty glossary. Raw start: {snippet!r}")
    else:
        clamp_scenes_to_blocks(context, blocks)
        enrich_scenes_with_block_text(context, blocks)
    return context


async def _attribute_scene(
    client: httpx.AsyncClient,
    scene: SceneHint,
    by_num: dict[int, SubtitleBlock],
    cfg: TranslationConfig,
    characters: list[CharacterHint],
) -> dict[int, str]:
    present = set(scene.participants)
    roster = "\n".join(
        f"- {h.source} ({gender_mark(h.gender) or '?'})"
        for h in characters if h.source in present
    )
    block_lines = [
        f"[{n}] {by_num[n].text.replace(chr(10), ' ')}"
        for n in range(scene.start, scene.end + 1) if n in by_num
    ]
    if not block_lines or not roster:
        return {}
    user_msg = build_attribution_user_message(roster, block_lines)
    try:
        cfg.calls.count("attribution")
        raw = await call_chat_api(
            client, ATTRIBUTION_SYSTEM_PROMPT, user_msg, cfg,
            max_tokens=len(block_lines) * 20 + 100,
        )
    except Exception as e:
        cfg.warn(f"    Attribution failed for blocks {scene.start}-{scene.end}: {e}")
        return {}
    return parse_attribution_response(raw, scene, characters)


async def refine_scene_attribution(
    client: httpx.AsyncClient,
    context: FileContext,
    blocks: list[SubtitleBlock],
    cfg: TranslationConfig,
) -> None:
    """Fill `SceneHint.attribution` for multi-block scenes with named
    participants. One small LLM call per target scene, bounded by concurrency."""
    targets = attribution_targets(context, cfg)
    if not targets:
        return
    by_num = {b.number: b for b in blocks}
    sem = asyncio.Semaphore(max(1, cfg.concurrency))

    async def do(scene: SceneHint) -> None:
        async with sem:
            scene.attribution = await _attribute_scene(
                client, scene, by_num, cfg, context.characters)

    await asyncio.gather(*(do(s) for s in targets))
